### PR TITLE
[WIP] KPI 

### PR DESCRIPTION
**Pull request by xanf**

### CODE REVIEW

File: app/models/kpi.rb [hunk 1]
```diff
@@ -0,0 +1,366 @@
+class Kpi < CouchRest::Model::Base
```
Comment (relledge) [inline app/models/kpi.rb]: Wouldn't a better model name be "KeyPerformanceIndicator"?   I had to ask what "KPI" meant.

File: app/models/kpi.rb [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+      [6, 14],
+      [15, 30],
+      [31, 92]
+    ]
+  }
+
+  ASSESSMENT_TYPE = {
+    name: 'assessment',
+    calculations: ['completed_asssessment', 'approved_assessment']
+  }
```
Comment (relledge) [inline app/models/kpi.rb]: Remove trailing whitespace

File: app/models/kpi.rb [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+      [15, 30],
+      [31, 92]
+    ]
+  }
+
+  ASSESSMENT_TYPE = {
+    name: 'assessment',
+    calculations: ['completed_asssessment', 'approved_assessment']
+  }	
+    CASE_ACTION_PLANNING_TYPE = {
```
Comment (relledge) [inline app/models/kpi.rb]: Indentation of CASE_ACTION_PLANNING_TYPE should align with ASSESSMENT_TYPE above

File: app/models/kpi.rb [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    calculations: [
+      'elapsed_time_from_opening_to_closure',
+      'elapsed_time_from_opening_to_closure_hight_risk',
+      'reason_for_case_closure',
+      'case_closure_rate'
+    ],
+    days_ranges: [
+      [0, 30],
+      [31, 92],
+      [93, 185],
```
Comment (relledge) [inline app/models/kpi.rb]: Remove comma after last element of array

File: app/models/kpi.rb [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    calculations: ['client_satisfaction']
+  }
+
+  OTHER_TYPE = {
+    name: 'other_kpi',
+    calculations: [
+      'case_load',
+      'supervisor_to_case_worker_ratio',
+      'caseworkers_active_on_gbvims_plus'
+    ],
```
Comment (relledge) [inline app/models/kpi.rb]: Remove comma after last element of hash

File: app/models/kpi.rb [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+  TYPES = [
+    INTRODUCTION_AND_ENGAGEMENT_TYPE,
+    ASSESSMENT_TYPE,
+    CASE_ACTION_PLANNING_TYPE,
+    IMPLEMENT_THE_CASE_ACTION_PLAN_TYPE,
+    CASE_FOLLOW_UP_TYPE,
+    CASE_CLOSURE_TYPE,
+    CLIENT_SATISFACTION_TYPE,
+    OTHER_TYPE,
```
Comment (relledge) [inline app/models/kpi.rb]: Remove comma after last element of array

File: app/models/kpi.rb [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+  ]
+
+  NEEDED_DATE_RANGE = ['follow_up_meetings', 'completed_asssessment']
+
+  def self.minimum_reportable_fields
```
Comment (relledge) [inline app/models/kpi.rb]: Prefer to have a  "class << self" block instead of having these "def self.<something" method definitions.

```suggestion
  class << self
    def minimum_reportable_fields
       <method lines>
    end
     
    def get_default_range_from(range)
       <method lines>
    end 
     ...  
  end
```

File: app/models/kpi.rb [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    {
+      'string' => [
+        INTRODUCTION_AND_ENGAGEMENT_TYPE[:name],
+        ASSESSMENT_TYPE[:name],
+        CASE_ACTION_PLANNING_TYPE[:name],
+        IMPLEMENT_THE_CASE_ACTION_PLAN_TYPE[:name],
+        CASE_FOLLOW_UP_TYPE[:name],
+        CASE_CLOSURE_TYPE[:name],
+        CLIENT_SATISFACTION_TYPE[:name],
+        OTHER_TYPE[:name],
```
Comment (relledge) [inline app/models/kpi.rb]: Remove comma after last element of array

File: app/models/kpi.rb [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    }
+  end
+
+  def self.get_default_range_from(range)
+    range.each_with_object({}) do |(from, to), default|
+      default["#{from}-#{to}"] = 0
+    end
+  end
+
+  def self.to_percent value
```
Comment (relledge) [inline app/models/kpi.rb]: Use () around parameter list

```suggestion
  def to_percent(value)
```

File: app/models/kpi.rb [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    "#{(value.to_f * 100).ceil(3)}%"
+  end
+
+  def self.format_result_for_reduced_view(view_result, range = [])
+    view_result.rows.each_with_object(get_default_range_from range) do |row, result|
```
Comment (relledge) [inline app/models/kpi.rb]: add () around parameters for method call "get_default_range_from"

File: app/models/kpi.rb [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+      result[row.key.kind_of?(Array) ? row.key[-1] : row.key] = row.value
+    end
+  end
+
+  def self.format_percents_from_rows(rows)
+    rows
+    .map {|key, result| [key, result.to_f / rows['_count'].to_f]} 
+    .to_h
+    .except '_count'
```
Comment (relledge) [inline app/models/kpi.rb]: would prefer this
```suggestion
    rows.map {|key, result| [key, result.to_f / rows['_count'].to_f]}.to_h.except('_count')
```

File: app/models/kpi.rb [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+      with :owned_by_agency, current_user[:organization]
+      with(:registration_date).between(
+        Date.parse(date_range[0])..Date.parse(date_range[1])
+      ) if date_range[0].present? && date_range[1].present?
+    end
+    all_active_cases_search.total || 1
+  end
+
+  def self.generate_kpi(types, current_user, date_range)
+    kpi_calculation_results = Parallel.map(types, in_threads: types.length) do |kpi_type|
```
Comment (relledge) [inline app/models/kpi.rb]: Remove trailing whitespace

File: app/models/kpi.rb [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+      ) if date_range[0].present? && date_range[1].present?
+    end
+    all_active_cases_search.total || 1
+  end
+
+  def self.generate_kpi(types, current_user, date_range)
+    kpi_calculation_results = Parallel.map(types, in_threads: types.length) do |kpi_type| 
+      [kpi_type, send(kpi_type, current_user, date_range)]
+    end
+    kpi_calculation_results.each_with_object({}) do |(kpi_type, calculation_result), kpi|
```
Comment (relledge) [inline app/models/kpi.rb]: Remove trailing whitespace

File: app/models/kpi.rb [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+    all_active_cases_search.total || 1
+  end
+
+  def self.generate_kpi(types, current_user, date_range)
+    kpi_calculation_results = Parallel.map(types, in_threads: types.length) do |kpi_type| 
+      [kpi_type, send(kpi_type, current_user, date_range)]
+    end
+    kpi_calculation_results.each_with_object({}) do |(kpi_type, calculation_result), kpi| 
+      group = TYPES.detect {|e| e[:calculations].include?(kpi_type) }
+      kpi[group[:name]] ||= {};
```
Comment (relledge) [inline app/models/kpi.rb]: Remove ;

File: app/models/kpi.rb [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+      kpi[group[:name]][kpi_type] = calculation_result
+    end
+  end
+
+  def self.rate_of_report(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: Parameter "date_range" isn't used.  If it is necessary, use _ at the beginning
```suggestion
  def self.rate_of_report(current_user, _date_range)
```

File: app/models/kpi.rb [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+      )
+      with :owned_by_agency, current_user[:organization]
+    end
+    { 
+      first_month: first_past_month_rate.total, 
+      second_month: second_past_month_rate.total
+    }
+  end
+
+  def self.elapsed_time_for_incident_and_report(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: Unused argument: date_range

File: app/models/kpi.rb [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+    format_result_for_reduced_view(
+      Incident.by_days_between_incident_and_report(
+        start_key: [current_user[:organization]],
+        end_key: [current_user[:organization], {}]
+      ).reduce.group_level(2),
+      INTRODUCTION_AND_ENGAGEMENT_TYPE[:days_ranges]
+    )
+  end
+
+  def self.average_delay_of_access(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+      with :owned_by_agency, current_user[:organization]
+      with(:registration_date).between(
+        Date.parse(date_range[0])..Date.parse(date_range[1])
+      ) if date_range[0].present? && date_range[1].present?
+      without :assessment_completion_timing, nil
+    end
+    completed_asssessment_cases.total.to_f / active_cases_count.to_f
+  end
+
+  def self.approved_assessment(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+    approved_asssessment_cases = Child.search do
+      with :child_status, Record::STATUS_OPEN
+      with :bia_approved, true
+      with :owned_by_agency, current_user[:organization]
+      without :assessment_completion_timing, nil
+    end
+    approved_asssessment_cases.total.to_f / active_cases_count.to_f
+  end
+
+  def self.needed_safety_plan(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+    cases_with_needed_and_completed_safety_plan = Child.search do
+      with :child_status, Record::STATUS_OPEN
+      with :safety_plan_needed, true
+      with :owned_by_agency, current_user[:organization]
+      without :safety_plan_completion_timing, nil
+    end
+    cases_with_needed_and_completed_safety_plan.total.to_f / active_cases_count.to_f
+  end
+
+  def self.completed_case_action_plan(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+    active_cases_count = get_all_open_cases_count(current_user)
+    cases_with_filled_action_plan = Child.search do
+      with :child_status, Record::STATUS_OPEN
+      with :owned_by_agency, current_user[:organization]
+      without :action_plan_section, nil
+    end
+    cases_with_filled_action_plan.total.to_f / active_cases_count.to_f
+  end
+
+  def self.approved_case_action_plan(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+    approved_cases_with_filled_action_plan = Child.search do
+      with :child_status, Record::STATUS_OPEN
+      with :case_plan_approved, true
+      with :owned_by_agency, current_user[:organization]
+      without :action_plan_section, nil
+    end
+    approved_cases_with_filled_action_plan.total.to_f / active_cases_count.to_f
+  end
+
+  def self.services_provided(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+      Child.by_service_provided(
+        start_key: [current_user[:organization]],
+        end_key: [current_user[:organization], {}]
+      ).reduce.group_level(2)
+    )
+  end
+
+  def self.follow_up_meetings(current_user, (from, to))
+    period = Date.parse(from)..Date.parse(to)
+    Child.by_followup_date(keys: period.to_a.map{|r|[current_user[:organization], r]}).all.count;
```
Comment (relledge) [inline app/models/kpi.rb]: Remove trailing ;

File: app/models/kpi.rb [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+  end
+
+  def self.average_follow_up_meetings(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+    all_active_cases_count = get_all_open_cases_count(current_user)
+    all_followups = Child.by_followup_date(
+      start_key: [current_user[:organization]],
+      end_key: [current_user[:organization], {}]
+    ).all;
+    all_followups.count.to_f / all_active_cases_count.to_f
+  end
+
+  def self.elapsed_time_from_opening_to_closure(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+        Child.by_days_between_case_opening_and_closure(
+          start_key: [current_user[:organization]],
+          end_key: [current_user[:organization], {}]
+        ).reduce.group_level(2),
+        CASE_CLOSURE_TYPE[:days_ranges]
+      )
+    )
+  end
+
+  def self.elapsed_time_from_opening_to_closure_hight_risk(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 27]
```diff
@@ -0,0 +1,366 @@
...
+        Child.by_days_between_case_opening_and_closure_high_risk(
+          start_key: [current_user[:organization]],
+          end_key: [current_user[:organization], {}]
+        ).reduce.group_level(2),
+        CASE_CLOSURE_TYPE[:days_ranges]
+      )
+    )
+  end
+
+  def self.case_closure_rate(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 28]
```diff
@@ -0,0 +1,366 @@
...
+    cases_closed_past_month = Child.search do
+      with :owned_by_agency, current_user[:organization]
+      with(:date_closure).between(
+        Date.today.at_beginning_of_month.prev_month..Date.today.end_of_month.prev_month
+      )
+    end
+    cases_closed_past_month.total
+  end
+
+  def self.reason_for_case_closure(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 29]
```diff
@@ -0,0 +1,366 @@
...
+      format_result_for_reduced_view(
+        Child.by_closure_reason(
+          start_key: [current_user[:organization]],
+          end_key: [current_user[:organization], {}]
+        ).reduce.group_level(2)
+      )
+    )
+  end
+
+  def self.client_satisfaction(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 30]
```diff
@@ -0,0 +1,366 @@
...
+    active_cases_count = get_all_open_cases_count(current_user)
+    cases_with_satisfied_survivor = Child.client_statisfied(key: current_user[:organization]).count
+    cases_with_satisfied_survivor.to_f / active_cases_count.to_f
+  end
+
+  def self.case_load(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 31]
```diff
@@ -0,0 +1,366 @@
...
+        res['30'] = res['30'].to_i + 1 
+      else
+        res['other'] = res['other'].to_i + 1 
+      end
+      res['_count'] = res['_count'].to_i + 1
+    end
+    format_percents_from_rows(statistic)
+  end
+
+  def self.supervisor_to_case_worker_ratio(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/kpi.rb [hunk 32]
```diff
@@ -0,0 +1,366 @@
...
+        supervisor_to_case_worker_in_group = caseworkers_in_group_count.to_f / group['role-gbv-case-management-supervisor'].to_f
+        type = supervisor_to_case_worker_in_group < 5 ? 'lt' : 'gte'
+        res[type] = res[type].to_i + 1
+        res['_count'] = res['_count'].to_i + 1
+      end
+    end
+    format_percents_from_rows(supervisor_to_case_worker)
+  end
+
+  def self.caseworkers_active_on_gbvims_plus(current_user, date_range)
```
Comment (relledge) [inline app/models/kpi.rb]: unused argument: date_range

File: app/models/child.rb [hunk 33]
```diff
@@ -165,6 +165,154 @@ def initialize *args
                     }
                  }
               }"
+
+    view :by_followup_date,
+    :map => "function(doc) {
```
Comment (relledge) [inline app/models/child.rb]: Use spaces instead of tabs for indent

File: app/models/child.rb [hunk 34]
```diff
@@ -94,7 +94,7 @@ def initialize *args
     view :by_date_of_birth
 
     view :by_name,
-         :map => "function(doc) {
+    :map => "function(doc) {
```
Comment (relledge) [inline app/models/child.rb]: Why was this changed?

File: app/models/report.rb [hunk 35]
```diff
@@ -34,6 +34,7 @@ class Report < CouchRest::Model::Base
   localize_properties [:name, :description]
   property :module_ids, [String]
   property :record_type #case, incident, etc.
+	property :data
```
Comment (relledge) [inline app/models/report.rb]: Use spaces instead of tab

File: app/models/report.rb [hunk 36]
```diff
@@ -135,6 +135,11 @@ def build_report
     sys = SystemSettings.current
     primary_range = sys.primary_age_range
     age_ranges = sys.age_ranges[primary_range]
+		
+		if self.data.present?
```
Comment (relledge) [inline app/models/report.rb]: Don't use tabs.
Remove trailing ; below

File: app/models/kpi.rb (hunk 1 above)
Comment (relledge) [inline app/models/kpi.rb]: Need rspec tests for kpi model

File: config/locales/es.yml [hunk 37]
```diff
@@ -832,6 +832,7 @@ es:
       case: "Caso"
       child: "Caso"
       incident: "Incidente"
+      kpi: "KPI"
```
Comment (relledge) [inline config/locales/es.yml]: The only locale yml you need to update is 'en.yml'.   We populate the others by loading en.yml into Transifex and then let other folks translate it for us.

File: app/controllers/reports_controller.rb [hunk 38]
```diff
@@ -77,6 +77,10 @@ def create
     end
 
     if (@report.valid?)
+      if @report.record_type == "kpi"
+        @report.group_dates_by = params[:date_from] != "" && params[:date_to] != "" ? [params[:date_from], params[:date_to]] : []
```
Comment (relledge) [inline app/controllers/reports_controller.rb]:
```suggestion
        @report.group_dates_by = params[:date_from].present? && params[:date_to].present? ? [params[:date_from], params[:date_to]] : []
```

File: app/controllers/reports_controller.rb [hunk 39]
```diff
@@ -92,7 +96,16 @@ def edit
 
   def update
     authorize! :update, @report
-
+    if params[:report][:record_type] == "kpi"
```
Comment (relledge) [inline app/controllers/reports_controller.rb]: This if block is very similar to the one above in 'create'.  Extract this common logic into a method, perhaps loaded in a before action.

File: app/assets/javascripts/views/reports_form.js [hunk 40]
```diff
@@ -32,6 +32,36 @@ _primero.Views.ReportForm = _primero.Views.Base.extend({
       $('#report_aggregate_by, #report_disaggregate_by, .report_filter_value_string_row select.report_filter_input, .report_filter_attribute').addClass('chosen-rtl');
     }
   },
+  load_lookups_and_disable_useless_fields: function (e) {
+		var reportSection = $("#report-section");
```
Comment (relledge) [inline app/assets/javascripts/views/reports_form.js]: Use spaces instead of tabs

File: cookbook/Berksfile.lock [hunk 41]
```diff
@@ -36,7 +36,6 @@ GRAPH
     yum (~> 3.0)
     yum-epel (>= 0.0.0)
     yum-erlang_solutions (>= 0.0.0)
-  java (1.22.0)
```
Comment (relledge) [inline cookbook/Berksfile.lock]: Why did this change?

File: db/users/roles.rb [hunk 42]
```diff
@@ -653,7 +653,7 @@ def create_or_update_role(role_hash)
 
 
 gbv_caseworker_forms = [
-  "action_plan_form", "gbv_case_closure_form", "consent_for_referrals", "other_documents", "record_owner",
+  "action_plan_form", "gbv_case_closure_form", "gbv_assessment", "gbv_introduction_and_engagement", "gbv_client_satisfaction", "gbv_case_action_planning", "gbv_case_closure", "gbv_other_kpi", "gbv_follow_up", "gbv_implement_case_action_planning", "consent_for_referrals", "other_documents", "record_owner",
```
Comment (relledge) [inline db/users/roles.rb]: This line is too long.

File: db/users/default_modules.rb [hunk 43]
```diff
@@ -250,9 +250,9 @@ def create_or_update_module(module_hash)
 create_or_update_module(
   name: "GBV",
   description: "Gender Based Violence",
-  associated_record_types: ["case", "incident"],
+  associated_record_types: ["case", "incident", "kpi"],
   associated_form_ids: [
-    "record_owner", "consent_for_services", "gbv_survivor_information",
+    "record_owner", "consent_for_services", "gbv_survivor_information", "gbv_assessment", "gbv_introduction_and_engagement", "gbv_client_satisfaction", "gbv_case_action_planning", "gbv_case_closure", "gbv_other_kpi", "gbv_follow_up", "gbv_implement_case_action_planning",
```
Comment (relledge) [inline db/users/default_modules.rb]: This line is too long

File: db/forms/kpi/gbv_assessment.rb [hunk 44]
```diff
@@ -0,0 +1,28 @@
+gbv_assessment_fields = [
+  Field.new({"name" => "completed_asssessment",
```
Comment (relledge) [inline db/forms/kpi/gbv_assessment.rb]:
```suggestion
  Field.new({"name" => "completed_assessment",
```

Comment (relledge) [inline db/forms/kpi/gbv_assessment.rb]: completed_assessment is misspelled throughout... if this changes, other references in the code will need to change

File: db/forms/kpi/gbv_assessment.rb [hunk 45]
```diff
@@ -0,0 +1,28 @@
...
+    "mobile_visible" => false,
+    "type" => "tick_box",
+    "display_name_en" => "Completed assessment",
+    "editable" => false,
```
Comment (relledge) [inline db/forms/kpi/gbv_assessment.rb]: All of your fields are set as "editable" false.  This prevents the field from being changed in the Form Editor.   Is that what you intended?

File: db/forms/kpi/gbv_assessment.rb [hunk 46]
```diff
@@ -0,0 +1,28 @@
...
+    "type" => "tick_box",
+    "display_name_en" => "Completed assessment, supervisor approved",
+    "editable" => false,
+    "disabled" => true
+  }),
+];
+
+FormSection.create_or_update_form_section({
+  unique_id: "gbv_assessment",
+  parent_form: "kpi",
```
Comment (relledge) [inline db/forms/kpi/gbv_assessment.rb]: I'm struggling to understand why it is necessary to add a new group of KPI forms.  How do these forms/fields get populated?  How are they used?   Should they just be properties on the kpi model instead?

File: app/models/incident.rb [hunk 47]
```diff
@@ -7,37 +7,37 @@ class Incident < CouchRest::Model::Base
     {
       "source" => ["survivor_code_no"],
       "target" => "survivor_code"
-    }, {
```
Comment (relledge) [inline app/models/incident.rb]: Why did these change?

File: app/models/kpi.rb [hunk 48]
```diff
@@ -0,0 +1,366 @@
...
+    CLIENT_SATISFACTION_TYPE,
+    OTHER_TYPE,
+  ]
+
+  NEEDED_DATE_RANGE = ['follow_up_meetings', 'completed_asssessment']
+
+  def self.minimum_reportable_fields
+    {
+      'string' => [
+        INTRODUCTION_AND_ENGAGEMENT_TYPE[:name],
```
Comment (relledge) [inline app/models/kpi.rb]: In other models (Child, Incident, etc) these fields are defined as properties either on the model itself or somewhere in one of the included model concerns.  It doesn't rely on the field being defined in a form somewhere.  Where are these fields defined?

File: app/models/kpi.rb (hunk 1 above)
Comment (relledge) [inline app/models/kpi.rb]: Kpi model should not derive from CouchRest::Model::Base if it isn't being stored in the database.  From what I see, this model is used to pull back stuff from other models (Child, Incident, User).  If that is the case, this model doesn't need to be a CouchRest model.

File: app/models/report.rb [hunk 49]
```diff
@@ -416,10 +421,12 @@ def pivots
   end
 
   def pivot_fields
-    @pivot_fields ||= Field.find_by_name(pivots)
-      .group_by{|f| f.name}
-      .map{|k,v| [k, v.first]}
-      .to_h
+    @pivot_fields ||= pivots.present? ?
```
Comment (relledge) [inline app/models/report.rb]: Is adding the pivots.present causing the pivots calculation to be executed twice here?   If so, I'd recommend improving that method to only execute once.  
 Example:
        @pivots ||= (self.aggregate_by || []) + (self.disaggregate_by || [])

File: app/models/report.rb [hunk 50]
```diff
@@ -416,10 +421,12 @@ def pivots
...
+			 Field.find_by_name(pivots)
```
Comment (relledge) [inline app/models/report.rb]: Don't use tabs

File: app/models/report.rb [hunk 51]
```diff
@@ -135,6 +135,11 @@ def build_report
     sys = SystemSettings.current
     primary_range = sys.primary_age_range
     age_ranges = sys.age_ranges[primary_range]
+		
+		if self.data.present?
+      self.data = { values: self.data };
```
Comment (relledge) [inline app/models/report.rb]: I'm concerned that this short-circuits the basic way reports work.  Here it seems the report data is being fetched from the DB when the report is created and saved on the report record, then just the same report data re-displayed over and over each time the user clicks to view the report.  Normally, a report, as saved in the db, is just info on how to retrieve the report data from solr.  In that scenario, fresh report data is fetched each time a user clicks on a report to view the report.

File: app/models/child.rb [hunk 52]
```diff
@@ -165,6 +165,154 @@ def initialize *args
...
+          var dateOpen = new Date(doc['registration_date']);
+          var dateClosure = new Date(doc['date_closure']);
+          var oneDay = 1000 * 60 * 60 * 24;
+          var dayDifference = +((dateClosure - dateOpen) / oneDay).toFixed(0)
+          var dayRange = #{Kpi::CASE_CLOSURE_TYPE[:days_ranges]}
+            .filter(function (ranges) {
+              return dayDifference >= ranges[0] && dayDifference <= ranges[1]
+            });
+          emit([doc['owned_by_agency'], dayRange[0] ? dayRange[0][0] + '-' + dayRange[0][1] : 'other'], null)
+          emit([doc['owned_by_agency'], '_count'], null)
```
Comment (relledge) [inline app/models/child.rb]: Not sure why you would want to emit a row in the view with the string '_count'.  I tested this.  For 1 case record, 2 rows appeared... 1 for the date range and the other with _count.   Seems odd.

File: app/models/child.rb [hunk 53]
```diff
@@ -165,6 +165,154 @@ def initialize *args
...
+        }
+      }",
+    :reduce => "function(key, values, rereduce) {
+        if (rereduce) {
+          return sum(values);
+        }
+        return values.length
+      }"
+
+    view :by_closure_reason,
```
Comment (relledge) [inline app/models/child.rb]: I think most if not all of these new views could be better implemented using the existing report logic which queries solr.  The existing report logic provides the capability to filter, group dates by a range, etc

File: db/forms/kpi/gbv_assessment.rb [hunk 54]
```diff
@@ -0,0 +1,28 @@
+gbv_assessment_fields = [
```
Comment (relledge) [inline db/forms/kpi/gbv_assessment.rb]: Is the only purpose of these new forms to drive the select options for row/column on the reports editor?  If so i think we need to do this a different way... perhaps hard coding the options

**Comment by pnabutovsky** [comment]
KPIs will be addressed in Primero v2.0
